core/src/main/scala/org/apache/spark/h2o/utils/H2OSchemaUtils.scala

```diff
@@ -84,69 +96,74 @@ object H2OSchemaUtils {
     df.map[Row] { row: Row =>
       val result = ArrayBuffer.fill[Any](numberOfColumns)(null)
       val fillBufferPartiallyApplied = fillBuffer(nameToIndexMap, result) _
-      originalSchema.fields.zipWithIndex.foreach { case (field, idx) =>
-        fillBufferPartiallyApplied(field, row(idx), None)
+      var idx = 0
+      val fields = originalSchema.fields
+      while (idx < fields.length) {
+        val field = fields(idx)
+        fillBufferPartiallyApplied(field.name, field.dataType, row(idx))
+        idx = idx + 1
       }
-      Row.fromSeq(result)
+      new GenericRowWithSchema(result.toArray, flatSchema)
     }
   }
 
   private def fillBuffer
-  (flatSchemaIndexes: Map[String, Int], buffer: ArrayBuffer[Any])
-  (field: StructField, data: Any, prefix: Option[String] = None): Unit = {
+      (flatSchemaIndexes: Map[String, Int], buffer: ArrayBuffer[Any])
+      (qualifiedName: String, dataType: DataType, data: Any) = {
     if (data != null) {
-      val StructField(name, dataType, _, _) = field
-      val qualifiedName = getQualifiedName(prefix, name)
       dataType match {
-        case BinaryType => fillArray(ByteType, flatSchemaIndexes, buffer, data, qualifiedName)
-        case MapType(_, valueType, _) => fillMap(valueType, flatSchemaIndexes, buffer, data, qualifiedName)
-        case ArrayType(elementType, _) => fillArray(elementType, flatSchemaIndexes, buffer, data, qualifiedName)
-        case StructType(fields) => fillStruct(fields, flatSchemaIndexes, buffer, data, qualifiedName)
+        case BinaryType =>
+          val binaryData = data.asInstanceOf[Array[Byte]].toSeq
```

Comment:
Here we are boxing array values. The question is whether we shouldn't duplicate `fillArray` method.